### PR TITLE
Introduce the capability to aggregate WebSocket continuation frames.

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
Users may find it beneficial to aggregate WebSocket continuation frames and their text or binary frame into a single frame instead of managing them individually.

Modifications:
- Introduce properties `WebSocketServiceBuilder.aggregateContinuation()` and `WebSocketClientBuilder.aggregateContinuation()` to enable the aggregation of continuation frames.

Result:
- You now have the option to aggregate WebSocket continuation frames for more streamlined handling.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java [hunk 1]
```diff
@@ -271,25 +277,47 @@ public void process(StreamDecoderInput in, StreamDecoderOutput<WebSocketFrame> o
...
                     } else {
-                        throw protocolViolation(WebSocketCloseStatus.INVALID_MESSAGE_TYPE,
-                                                "Cannot decode a web socket frame with opcode: " + frameOpcode);
+                        assert frameOpcode == WebSocketFrameType.CONTINUATION.opcode();
+                        decodedFrame = WebSocketFrame.ofPooledContinuation(payloadBuffer, finalFragment);
                     }
                     logger.trace("{} is decoded.", decodedFrame);
+
                     if (finalFragment) {
                         fragmentedFramesCount = 0;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java:299]: Side-note) This comment is not related to this PR. I don't see we need to track `fragmentedFramesCount` based on the current usage. a boolean flag seems enough to know fragmented frames are received.

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java [hunk 2]
```diff
@@ -271,25 +277,47 @@ public void process(StreamDecoderInput in, StreamDecoderOutput<WebSocketFrame> o
...
+                        if (aggregatingFrames.isEmpty()) {
+                            out.add(decodedFrame);
+                        } else {
+                            aggregatingFrames.add(decodedFrame);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java:303]: From my understanding, `maxFramePayloadLength` does not limit the total length of `aggregatingFrames`. I guess it could be a vulnerable point. 

What do you think of applying `maxFramePayloadLength` to the `aggregatingFrames` if `aggregateContinuation` is enabled? If reasonable, it is also worth noting on the Javadoc of `aggregateContinuation` option.

**Review by ikhoon — COMMENTED** [review]
This is cut out for me to implement #5167 😆

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java (hunk 1 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java:299]: That is correct. When I forked the original code, I also noticed that `fragmentedFramesCount` can be a boolean field but I didn't change it because
- There's no not so much performance penalty
- Keeping the original code as much as possible makes it simpler to update the forked source code later if there are changes in the upstream code.

https://github.com/netty/netty/blob/e8df52e442629214e0355528c00e873e213f0139/codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocket08FrameDecoder.java#L98

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java (hunk 2 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java:303]: That makes sense and is a good suggestion. 👍

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java:299]: Understood. It would be better to send a PR first before modifying it. Let's keep it as is for now.

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java [hunk 3]
```diff
@@ -271,25 +278,58 @@ public void process(StreamDecoderInput in, StreamDecoderOutput<WebSocketFrame> o
...
+                        }
                     } else {
                         fragmentedFramesCount++;
+                        if (aggregateContinuation) {
+                            aggregatingFramesLength += framePayloadLength;
+                            if (aggregatingFramesLength > maxFramePayloadLength) {
+                                throw protocolViolation(
+                                        WebSocketCloseStatus.MESSAGE_TOO_BIG,
+                                        // The message must not exceed 125 length:
+                                        // https://datatracker.ietf.org/doc/html/rfc6455/#section-5.5
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java:324]: nit; this comment probably isn't related since it refers to control frames only

```suggestion
```

File: core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java [hunk 4]
```diff
@@ -80,6 +81,7 @@ public final class WebSocketClientBuilder extends AbstractWebClientBuilder {
     private int maxFramePayloadLength = DEFAULT_MAX_FRAME_PAYLOAD_LENGTH;
     private boolean allowMaskMismatch;
     private List<String> subprotocols = ImmutableList.of();
+    private boolean aggregateContinuation;
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java:84]: minor question) Maybe it is just me, but I feel like most users won't be interested in continuation frames and `true` would be the more sensible default.

**Review by jrhee17 — APPROVED** [review]
Changes look straightforward enough 👍  Thanks @minwoox 🙇 👍 🙇

File: core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java (hunk 4 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java:84]: I think we may provide a higher API that can automatically aggregate continuation frames with `StreamMessage<String>` or `StreamMessage<ByteBuf>`.

File: core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java (hunk 3 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java]: The message is set to the reason phrase of the close frame and the close frame is one of control frames so I left the comment here. 😉 
However, I've changed to logic to truncate the message if it exceeds the limit in `WebSocketUtil` so I've moved this comment to there. PTAL. 😉

File: core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java (hunk 4 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java:84]: > most users won't be interested in continuation frame

I completely agree with your perspective on this matter, but I believe that setting true for `aggregateContinuation` as a default is a separate consideration.
If continuation frames are used that the server developers do not notice, the server may experience memory pressure. So I prefer to set it false. 🤔

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java:84]: > I think we may provide a higher API that can automatically aggregate continuation frames with StreamMessage<String> or StreamMessage<(ByteBuf|byte[])>.

Could you share an example of the usage? I'm not sure how it's going to be used because what we need here is the stream message of `WebSocketFrame` instead of strings or byte arrays. 🤔

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java:84]: Continuation is a low-level detail that many users might not be interested in for their business logic. I think we can abstract a set of WebSocket frames followed by continuation frames with `StreamMessage` so that users easily aggregate the whole message or subscribe to the messages one by one.

```java
TextStreamMessage extends StreamMessage<String> {
   CompletableFuture<String> aggregate();
}

ByteStreamMessage extends StreamMessage<byte[]> {
   CompletableFuture<byte[]> aggregate();
}

interface WebSocketStreamHandler {
   void onText(RequestContext ctx, TextStreamMessage texts, WebSocketWriter out);
   
   void onBinary(RequestContext ctx, TextStreamMessage bytes, WebSocketWriter out);
   ...  
}
```

```java
class EchoWebSocketStreamHandler implements WebSocketStreamHandler {
   void onText(RequestContext ctx, TextStreamMessage texts, WebSocketWriter out) {
     // If users want to get the whole text
     texts.aggregate().thenAccept(text -> {
        out.write(text);
     });
     // To get a text frame one by one with backpressure
     texts.subscribe(...)
   }
   
   void onBinary(RequestContext ctx, TextStreamMessage bytes, WebSocketWriter out);
}
```

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketClientBuilder.java:84]: Agreed. We don't know which one is preferred by users. Both versions could be provided later if necessary.

**Review by ikhoon — APPROVED** [review]
This is a useful option! 🚀🙇‍♂️
